File: 백준/Silver/1699. 제곱수의 합/제곱수의 합.java
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int maxN = 100004;
        int[] dp = new int[maxN];

        int n = scanner.nextInt();

        for (int i = 1; i <= n; i++) {
            dp[i] = i;
        }

        for (int i = 1; i <= n; i++) {
            for (int divNum = 1; divNum * divNum <= i; divNum++) {
                dp[i] = Math.min(dp[i], dp[i - divNum * divNum] + 1);
            }
        }

        System.out.println(dp[n]);
        scanner.close();
    }
}

File: 백준/Silver/1699. 제곱수의 합/제곱수의 합.cc
#include<iostream>
#include<queue>
#include <utility>
using namespace std;

const int maxN = 100004;
int n;
int dp[maxN];

int main() {

	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> n;
	
	for (int i = 1; i <= n; i++) {
		dp[i] = i;
	}


	for (int i = 1; i <= n; i++) {
		for (int divNum = 1; divNum * divNum <= i; divNum++) {
			dp[i] = min(dp[i], dp[i - divNum * divNum] + 1);
		}

	}
	cout << dp[n];

	return 0;
}